[FuncSpec] Enable SpecializeLiteralConstant by default

hazzlim opened the pull request:
Enable specialization on literal constant arguments by default in Function Specialization.

hazzlim committed:
[FuncSpec] Enable SpecializeLiteralConstant by default

Enable specialization on literal constant arguments by default in
Function Specialization.

hazzlim commented:
Change in compile-times for CTMark (measured as instruction-count) from applying this patch, measured on an AArch64 (Graviton 3) machine:

O3 (Non-LTO):

```
Benchmark        |  Change  |
-----------------+----------+
ClamAV           |   0.00%  |
7zip             |  +0.01%  |
kimwitu++        |  -0.01%  |
tramp3d-v4       |   0.00%  |
sqlite3          |  +0.02%  |
mafft            |   0.00%  |
SPASS            |   0.00%  |
lencod           |   0.00%  |
consumer-typeset |  +0.03%  |
Bullet           |   0.00%  |
-----------------+----------+
GEOMEAN          |   0.00%  |
```

ReleaseLTO-g:

```
Benchmark        |  Change  |
-----------------+----------+
ClamAV           |  +0.01%  |
7zip             |  -0.01%  |
kimwitu++        |   0.00%  |
tramp3d-v4       |  +0.01%  |
sqlite3          |  +0.01%  |
mafft            |   0.00%  |
SPASS            |  -0.01%  |
lencod           |  +0.02%  |
consumer-typeset |  +0.05%  |
Bullet           |  -0.01%  |
-----------------+----------+
GEOMEAN          |  +0.01%  |
```

david-arm commented on llvm/test/Transforms/FunctionSpecialization/literal-const.ll:
```diff
@@ -1,7 +1,8 @@
 ; RUN:        -force-specialization < %s | FileCheck %s -check-prefix CHECK-NOLIT
 ; RUN: opt -S --passes="ipsccp<func-spec>" \
-; RUN:        -funcspec-for-literal-constant \
+; RUN:        -funcspec-for-literal-constant=true \
```
Is it worth removing the `-funcspec-for-literal-constant=true` to defend this being enabled by default?

hazzlim replied:
Good point - removed :)

hazzlim committed:
Remove redundant funcspec-for-literal-constant=true in lit test

ChuanqiXu9 reviewed (commented):
LG. We did similar thing in downstream.

hazzlim committed:
Address review comments 1

- Improve comment regarding multiple iterations

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -682,10 +679,9 @@ bool FunctionSpecializer::run() {
-    if (!Inserted && !Metrics.isRecursive && !SpecializeLiteralConstant)
+    // When specialization on literal constants is disabled, only consider
+    // recursive functions when running multiple times.
+    if (!SpecializeLiteralConstant && !Inserted && !Metrics.isRecursive)
```
Ah yep thank you for clarifying this point, I have updated the comment to be clearer on this point. I reordered the checks because it seemed to make more sense to check SpecializeLiteralConstant=false, given that it is only in this case that we limit multiple iterations to recursive functions.

On this point, I was actually thinking - is it not hypothetically possible that a specialized function could also be found to return a constant _pointer_, e.g. a function pointer?

labrinea replied:
Hmm, that did not occur to me before good point. Would it be too much to ask adding a regression test? :) We already have one for tracking return values, perhaps you could use that as a template?
https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/FunctionSpecialization/track-return.ll

hazzlim replied:
No problem - I've added https://github.com/llvm/llvm-project/pull/113442/commits/4533583136a11f260b391c0bf9be2aeb72b389fc#diff-43f6ba03f40b792b303a3a0a05b40dccb7fe9898a649e93d71038be1ee6ebb0c

I've intentionally written it as a motivating case for removing this condition altogether, as it is an example of a non-recursive function that could be specialized on a successive iteration when `SpecializeConstantLiteral=false`.

Do you think that this makes? I don't believe it will have a significant impact on compile times - you could perhaps argue that my example is a tad contrived, however...

hazzlim committed:
Add regression test for pointer return value specialization

- Add regression test which motivates removing the conditional check to
  only consider recursive functions on successive iterations when
  SpecializeLiteralConstant=false

labrinea commented on llvm/test/Transforms/FunctionSpecialization/function-specialization-track-ptr-return.ll:
```diff
@@ -0,0 +1,147 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --include-generated-funcs --version 5
```
Can you remove the "function-specialization-" prefix from the file name? It is already under a folder named accordingly.

hazzlim replied:
Ah yep good point, Done.

labrinea commented on llvm/test/Transforms/FunctionSpecialization/function-specialization-track-ptr-return.ll:
```diff
@@ -0,0 +1,147 @@
+
+define internal i64 @compute(ptr %binop) {
+entry:
+  %res = call i64 %binop(i64 1, i64 1)
```
Nit picking: the naming of the called operand here doesn't seem right. Both plus and minus functions take one operand so they are unary. But here we are (accidentaly?) passing two.

hazzlim replied:
Sorry yes this was accidental - fixed.

labrinea commented on llvm/test/Transforms/FunctionSpecialization/function-specialization-track-ptr-return.ll:
```diff
@@ -0,0 +1,147 @@
+; CHECK-NEXT:    [[BINOP2:%.*]] = call ptr @select_binop.specialized.2(ptr @global_false)
+; CHECK-NEXT:    [[C1:%.*]] = call i64 @compute.specialized.3(ptr @plus)
+; CHECK-NEXT:    [[C2:%.*]] = call i64 @compute.specialized.4(ptr @minus)
+; CHECK-NEXT:    [[ADD:%.*]] = add i64 [[C1]], [[C2]]
```
I am puzzled why the lattice value is not updated here. I'd think the entire main function can be folded to the constant value 2.

hazzlim replied:
Hmm yes good point - I'll see if I can figure out what is going on here.

labrinea replied:
From a quick debug I noticed that plus and minus are not considered for specialization because they have not yet replaced the binop in the body of compute.specialized.3 and compute.specialized.4 (the Solver only knowns their lattice value), therefore findSpecializations does not find these callsites in the users list.

hazzlim replied:
Oh I see - I thought you meant that you would expect IPSCCP to find that plus and minus have constant return values, because they are only called once with a constant argument, but you meant you'd expect them to be specialized.

I think similarly IPSCCP doesn't handle propagating constants into function calls where the callee been found constant by the Solver. If you take the example IR:
```
define i64 @test() {
  %op = select i1 true, ptr @plus_test, ptr @minus_test
  %res = call i64 %op(i64 1)
  %sum = add i64 %res, 1
  ret i64 %sum
}
```
Running this through IPSCCP spits out:
```
`define i64 @test() {
  %res = call i64 @plus_test(i64 1)
  %sum = add i64 %res, 1
  ret i64 %sum
}

define internal i64 @plus_test(i64 %x) {
entry:
  %sum = add i64 %x, 1
  ret i64 %sum
}

define internal i64 @minus_test(i64 %x) {
entry:
  %diff = sub i64 %x, 1
  ret i64 %diff
}
`
define internal i64 @plus_test(i64 %x) {
entry:
  %sum = add i64 %x, 1
  ret i64 %sum
}

define internal i64 @minus_test(i64 %x) {
entry:
  %diff = sub i64 %x, 1
  ret i64 %diff
}
```

hazzlim replied:
I wonder if you could maybe teach `SCCPInstVisitor::visitCallBase` to handle indirect callees, although I'm not sure!

labrinea replied:
I meant that SCCP is designed such that first it "solves" (analyzes the program, caches lattice values to its data structures) and then it replaces uses of values which are known to be constant. Things happen in this order by design. If you replaced the users earlier without invalidating the internal data structures (like resetLatticeValue does) then they would contain stale data and crash. I was fighting with such bugs when I merged FuncSpec with SCCP.

labrinea replied:
I think it would be worth writing this finding in the header file of FuncSpec, in a new section under "ideas", perhaps named "limitations". Specifically not being able to consider specializations for callsites of indirect functions whose pointer operand has a lattice value that is known to be constant from previous iterations of FuncSpec. That is because SCCP has not yet replaced their uses.

hazzlim replied:
Thank you for clarifying :) I've added a comment to the header under `Limitations`, with the additional detail that the limitation also applies to indirect call-sites found constant by IPSCCP itself as well as previous FuncSpec iterations.

hazzlim committed:
Clean up regression test

- Rename regression test to remove redundant function-specialization
  prefix
- Correct `binop` -> `op` and remove superfluous parameter

hazzlim committed:
Add Limitations comment to header

labrinea committed:
Update FunctionSpecialization.h

nit picking on a comment

labrinea reviewed (approved):
Thanks for addressing all the raised issues! LGTM
